File: compose-ui/src/commonMain/kotlin/dev/dimension/flare/ui/component/TimelinePlaybackCoordinator.kt
package dev.dimension.flare.ui.component

import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.SideEffect
import androidx.compose.runtime.compositionLocalOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.layout.boundsInWindow
import androidx.compose.ui.layout.onGloballyPositioned
import androidx.compose.ui.layout.positionInWindow
import androidx.compose.ui.unit.toSize
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlin.math.abs

internal val LocalTimelinePlayback = compositionLocalOf<TimelinePlaybackCoordinator?> { null }
internal val LocalTimelineCarouselItem = compositionLocalOf<TimelineCarouselItem?> { null }

internal data class TimelineCarouselItem(
    val groupId: Any,
    val selected: Boolean,
    val isCarousel: Boolean = true,
)

internal class TimelinePlaybackCoordinator(
    val scope: CoroutineScope,
    private val arbiter: VideoPlaybackArbiter = VideoPlaybackArbiter.shared,
    private val memory: MediaPlaybackMemory = MediaPlaybackMemory.shared,
) {
    private val policy = TimelineAutoplayPolicy()
    private val candidates = linkedMapOf<Any, TimelineAutoplayPolicy.Candidate>()
    private val players = mutableMapOf<Any, (Boolean) -> Unit>()
    private val scrollingSources = mutableSetOf<Any>()
    private var selectionJob: Job? = null
    private var playingId: Any? = null
    private var closed = false
    private var suspended = false
    private var isPresentation = false
    private var immediateReturn = false
    var handoffUri: String? = null
        private set
    val mediaSelections = TimelineMediaSelections()
    private var viewerSelection: Pair<List<String>, String>? = null
    private val cleanup = mutableMapOf<Any, () -> Unit>()
    var viewport: Rect = Rect.Zero

    init {
        arbiter.register(
            this,
            ::stopCurrentPlayer,
            ::scheduleSelection,
            mediaSelections::returned,
            resume = {
                immediateReturn = true
                scheduleSelection()
            },
            willHandoff = { handoffUri = it },
        )
    }

    fun position(key: String): Double = memory.position(key)

    fun savePosition(
        key: String,
        seconds: Double,
    ) {
        memory.save(key, seconds)
    }

    fun selectMedia(
        groupId: Any,
        uri: String,
        userInitiated: Boolean,
    ) {
        if (userInitiated) {
            immediateReturn = false
            arbiter.interacted(this)
        }
        policy.returnedToMedia(groupId, uri)
        scheduleSelection()
    }

    fun selectViewerMedia(
        urls: List<String>,
        selectedUri: String?,
    ) {
        val selection = selectedUri?.takeIf { it in urls }?.let { urls.toList() to it }
        if (viewerSelection == selection) return
        viewerSelection = selection
        if (isPresentation) scheduleSelection()
    }

    fun present() {
        if (isPresentation) return
        isPresentation = true
        immediateReturn = true
        arbiter.present(this, viewerSelection?.second)
        scheduleSelection()
    }

    fun onClose(
        key: Any,
        action: () -> Unit,
    ) {
        cleanup[key] = action
    }

    fun register(
        id: Any,
        playback: (Boolean) -> Unit,
    ) {
        players[id] = playback
    }

    fun update(candidate: TimelineAutoplayPolicy.Candidate) {
        if (candidates[candidate.id] == candidate) return
        candidates[candidate.id] = candidate
        reconcile(allowStart = false)
        scheduleSelection()
    }

    fun remove(id: Any) {
        candidates.remove(id)
        reconcile(allowStart = false)
        players.remove(id)
        scheduleSelection()
    }

    fun setScrolling(
        source: Any,
        scrolling: Boolean,
        vertical: Boolean,
    ) {
        if (scrolling) {
            immediateReturn = false
            if (scrollingSources.add(source)) {
                arbiter.interacted(this)
                if (vertical) policy.verticalScrollBegan() else policy.interactWithCarousel(source)
            }
            selectionJob?.cancel()
        } else {
            scrollingSources.remove(source)
            scheduleSelection()
        }
    }

    fun close() {
        closed = true
        selectionJob?.cancel()
        arbiter.remove(this, viewerSelection?.first.orEmpty(), viewerSelection?.second)
        stopCurrentPlayer()
        players.clear()
        candidates.clear()
        cleanup.values.toList().forEach { it() }
        cleanup.clear()
    }

    fun setSuspended(value: Boolean) {
        if (suspended == value) return
        suspended = value
        if (value) {
            selectionJob?.cancel()
            stopCurrentPlayer()
            arbiter.release(this)
        } else {
            scheduleSelection()
        }
    }

    private fun scheduleSelection() {
        selectionJob?.cancel()
        if (closed || suspended || scrollingSources.isNotEmpty()) return
        if (immediateReturn) {
            reconcile(allowStart = true)
            if (playingId != null) immediateReturn = false
            return
        }
        selectionJob =
            scope.launch {
                delay(200)
                reconcile(allowStart = true)
            }
    }

    private fun reconcile(allowStart: Boolean) {
        if (closed || suspended) return
        // Scrolling only invalidates the current item; full selection runs at idle.
        if (!allowStart || scrollingSources.isNotEmpty()) {
            if (playingId != null && candidates[playingId]?.visible != true) {
                stopCurrentPlayer()
                arbiter.release(this)
            }
            return
        }
        val selectedUri = viewerSelection?.second.takeIf { isPresentation }
        val eligible = if (selectedUri == null) candidates.values else candidates.values.filter { it.mediaUri == selectedUri }
        val next = policy.select(eligible, scrolling = false)
        if (next == null) arbiter.settledWithoutVideo(this)
        if (next == playingId) return
        if (next != null && !arbiter.acquire(this)) {
            stopCurrentPlayer()
            return
        }
        playingId?.let { players[it]?.invoke(false) }
        playingId = next
        next?.let {
            immediateReturn = false
            players[it]?.invoke(true)
        }
        if (next == null) arbiter.release(this)
    }

    private fun stopCurrentPlayer() {
        val old = playingId
        playingId = null
        old?.let { players[it]?.invoke(false) }
        handoffUri = null
        policy.select(emptyList(), scrolling = true)
    }
}

/** Keeps background timelines paused while any page of the media viewer is open. */
@Composable
public fun MediaViewerPlayback(content: @Composable () -> Unit) {
    val scope = rememberCoroutineScope()
    val playback = remember(scope) { TimelinePlaybackCoordinator(scope) }
    CompositionLocalProvider(LocalTimelinePlayback provides playback, content = content)
    // Selection from the child is committed before preempting the timeline, so
    // the outgoing binding knows whether the same media is being handed over.
    SideEffect { playback.present() }
    DisposableEffect(playback) {
        onDispose { playback.close() }
    }
}

/** Records the last selected page for the timeline that opened this viewer. */
@Composable
public fun MediaViewerSelection(
    mediaUrls: List<String>,
    selectedUri: String?,
) {
    val playback = LocalTimelinePlayback.current
    SideEffect { playback?.selectViewerMedia(mediaUrls, selectedUri) }
}

@Composable
internal fun rememberTimelinePlayback(): TimelinePlaybackCoordinator {
    val inherited = LocalTimelinePlayback.current
    val scope = rememberCoroutineScope()
    val playback = inherited ?: remember(scope) { TimelinePlaybackCoordinator(scope) }
    DisposableEffect(playback, inherited) {
        onDispose { if (inherited == null) playback.close() }
    }
    return playback
}

private class VideoGeometry {
    var bounds: Rect = Rect.Zero
    var visible: Rect = Rect.Zero
}

@Composable
internal fun Modifier.timelineVideoAutoplay(
    playback: TimelinePlaybackCoordinator,
    id: Any,
    enabled: Boolean,
    mediaUri: String,
): Modifier {
    val item = LocalTimelineCarouselItem.current
    val geometry = remember(id) { VideoGeometry() }

    fun update() {
        val visible = if (playback.viewport.isEmpty) geometry.visible else geometry.visible.intersect(playback.viewport)
        playback.update(
            TimelineAutoplayPolicy.Candidate(
                id = id,
                groupId = item?.groupId,
                visible = enabled && !visible.isEmpty,
                selected = item?.selected ?: true,
                canStart = enabled && (item?.isCarousel != true || visible.width >= geometry.bounds.width * 0.6f),
                distance = abs(geometry.bounds.center.y - playback.viewport.center.y),
                mediaUri = mediaUri,
            ),
        )
    }
    SideEffect { update() }
    return onGloballyPositioned {
        geometry.bounds = Rect(it.positionInWindow(), it.size.toSize())
        geometry.visible = it.boundsInWindow()
        update()
    }
}

File: appleApp/Shared/FlareAppleCore/Sources/FlareAppleUI/TimelinePlaybackCoordinator.swift
import SwiftUI

struct TimelineScrollPhaseModifier: ViewModifier {
    let onChange: (Bool) -> Void

    func body(content: Content) -> some View {
        if #available(iOS 18.0, macOS 15.0, *) {
            content.onScrollPhaseChange { _, phase in onChange(phase != .idle) }
        } else {
            content
        }
    }
}

@MainActor
final class TimelinePlaybackCoordinator {
    private let arbiter: VideoPlaybackArbiter
    private var policy = TimelineAutoplayPolicy()
    private var candidates: [String: TimelineAutoplayPolicy.Candidate] = [:]
    private var players: [String: (Bool) -> Void] = [:]
    private var scrollingSources: Set<String> = []
    private var motionTasks: [String: Task<Void, Never>] = [:]
    private var selectionTask: Task<Void, Never>?
    private var playingID: String?
    private var returningGroupID: String?
    let mediaSelections = TimelineMediaSelections()
    private var suspended = false
    private var immediateReturn = false

    init(arbiter: VideoPlaybackArbiter = .shared) {
        self.arbiter = arbiter
        arbiter.register(self, stop: { [weak self] in self?.stopCurrentPlayer() },
                         reconsider: { [weak self] in self?.scheduleSelection() },
                         mediaReturned: { [weak self] urls, selected in
                             self?.mediaSelections.returned(urls: urls, selectedURL: selected)
                         }, resume: { [weak self] in
                             self?.immediateReturn = true
                             self?.scheduleSelection()
                         }, willHandoff: { VideoPlaybackSession.continuePlayback(to: $0) })
    }

    func register(id: String, playback: @escaping (Bool) -> Void) {
        players[id] = playback
    }

    func update(_ candidate: TimelineAutoplayPolicy.Candidate) {
        guard candidates[candidate.id] != candidate else { return }
        candidates[candidate.id] = candidate
        reconcile(allowStart: false)
        scheduleSelection()
    }

    func remove(id: String) {
        candidates.removeValue(forKey: id)
        reconcile(allowStart: false)
        players.removeValue(forKey: id)
        scheduleSelection()
    }

    func setScrolling(_ scrolling: Bool, source: String, vertical: Bool) {
        if scrolling {
            immediateReturn = false
            let motionSource = "motion:\(source)"
            motionTasks.removeValue(forKey: motionSource)?.cancel()
            scrollingSources.remove(motionSource)
            if scrollingSources.insert(source).inserted {
                returningGroupID = nil
                arbiter.interacted(self)
                if vertical { policy.verticalScrollBegan() }
                else { policy.interactedWithCarousel(source) }
            }
            selectionTask?.cancel()
        } else {
            scrollingSources.remove(source)
            scheduleSelection()
        }
    }

    // Geometry also covers macOS 14 / iOS 17, which have no scroll-phase API,
    // and programmatic scrolling. Real drag phases keep the source held longer.
    func moved(source: String, vertical: Bool) {
        guard !scrollingSources.contains(source) else { return }
        if !vertical, returningGroupID == source, immediateReturn {
            // Restoring the viewer's selected page is an immediate jump. Actual
            // drag/animation phases above still hold playback until scrolling ends.
            scheduleSelection()
            return
        }
        let motionSource = "motion:\(source)"
        if !scrollingSources.contains(source), !scrollingSources.contains(motionSource) {
            if vertical || returningGroupID != source {
                arbiter.interacted(self)
                if vertical { policy.verticalScrollBegan() }
                else { policy.interactedWithCarousel(source) }
            }
        }
        scrollingSources.insert(motionSource)
        selectionTask?.cancel()
        motionTasks[motionSource]?.cancel()
        motionTasks[motionSource] = Task { [weak self] in
            do { try await Task.sleep(for: .milliseconds(200)) } catch { return }
            guard let self else { return }
            self.scrollingSources.remove(motionSource)
            self.motionTasks.removeValue(forKey: motionSource)
            self.reconcile(allowStart: self.scrollingSources.isEmpty)
            if self.returningGroupID == source { self.returningGroupID = nil }
        }
    }

    func setSuspended(_ value: Bool) {
        suspended = value
        if value {
            selectionTask?.cancel()
            motionTasks.values.forEach { $0.cancel() }
            motionTasks.removeAll()
            scrollingSources.removeAll()
            switchPlayer(to: nil)
            arbiter.withdraw(self)
            _ = policy.select(from: [], isScrolling: true)
        } else {
            scheduleSelection()
        }
    }

    func selectMedia(groupID: String, mediaURL: String, userInitiated: Bool) {
        returningGroupID = userInitiated ? nil : groupID
        if userInitiated {
            immediateReturn = false
            arbiter.interacted(self)
        }
        policy.returnedToMedia(groupID: groupID, mediaURL: mediaURL)
        scheduleSelection()
    }

    private func scheduleSelection() {
        selectionTask?.cancel()
        guard scrollingSources.isEmpty, !suspended else { return }
        if immediateReturn {
            reconcile(allowStart: true)
            if playingID != nil { immediateReturn = false }
            return
        }
        selectionTask = Task { [weak self] in
            do { try await Task.sleep(for: .milliseconds(200)) } catch { return }
            self?.reconcile(allowStart: true)
            self?.returningGroupID = nil
        }
    }

    private func reconcile(allowStart: Bool) {
        guard !suspended else { return }
        // Geometry updates only need to validate the current video. Select from
        // the complete candidate set once scrolling has settled.
        guard allowStart, scrollingSources.isEmpty else {
            if let playingID, candidates[playingID]?.isVisible != true { switchPlayer(to: nil) }
            return
        }
        let next = policy.select(from: Array(candidates.values),
                                 isScrolling: false)
        if next == nil { arbiter.settledWithoutVideo(self) }
        switchPlayer(to: next)
    }

    private func switchPlayer(to next: String?) {
        guard next != playingID else { return }
        if next != nil, !arbiter.acquire(self) {
            stopCurrentPlayer()
            return
        }
        stopCurrentPlayer(resetPolicy: next == nil)
        playingID = next
        if let next {
            immediateReturn = false
            players[next]?(true)
        }
        else { arbiter.release(self) }
    }

    private func stopCurrentPlayer(resetPolicy: Bool = true) {
        let old = playingID
        playingID = nil
        if let old { players[old]?(false) }
        if resetPolicy { _ = policy.select(from: [], isScrolling: true) }
    }
}

private struct TimelinePlaybackCoordinatorKey: EnvironmentKey {
    static let defaultValue: TimelinePlaybackCoordinator? = nil
}

struct TimelineCarouselItem: Equatable {
    let groupID: String
    let isSelected: Bool
    var isCarousel = true
}

private struct TimelineCarouselItemKey: EnvironmentKey {
    static let defaultValue: TimelineCarouselItem? = nil
}

extension EnvironmentValues {
    var timelinePlaybackCoordinator: TimelinePlaybackCoordinator? {
        get { self[TimelinePlaybackCoordinatorKey.self] }
        set { self[TimelinePlaybackCoordinatorKey.self] = newValue }
    }

    var timelineCarouselItem: TimelineCarouselItem? {
        get { self[TimelineCarouselItemKey.self] }
        set { self[TimelineCarouselItemKey.self] = newValue }
    }
}
